src/odemis/acq/millmng.py

```diff
@@ -0,0 +1,504 @@
+
+        return stage_time
+
+    def estimate_drift_time(self, setting):
+        """
+
+        return (float): estimated time to acquire 1 anchor area
+        """
+        # nb_anchor_scanning = numpy.ceil(
+        #     self._settings[setting_nb]["duration"] / self._settings[setting_nb]["dcPeriod"])  # class , va's
+        drift_acq_time = 0
+        if setting.dcRoi.value != UNDEFINED_ROI:
+            nb_anchor_scanning = numpy.ceil(setting.duration.value / setting.dcPeriod.value)  # repeat?
+            drift_estimation = AnchoredEstimator(self._scanner, self._detector,
+                                                 setting.dcRoi.value, setting.dcDwellTime.value, max_pixels=512 ** 2, follow_drift=False)
+            drift_acq_time = drift_estimation.estimateAcquisitionTime() * nb_anchor_scanning
+
+        return drift_acq_time
+
+    def run(self):
+        """
+        The main function of the task class, which will be called by the future asynchronously
+
+        """
+        if not self._future:
+            return
+        self._future._task_state = RUNNING
+
+        try:
+            actual_time_per_site = None
+            # if no site is provided
+            if len(self.sites) == 0:
+                logging.debug("Site location is not provided by the user, cancelling the milling")
+                self._future._task_state = CANCELLED
+                raise CancelledError()
+            for site_idx, site in enumerate(self.sites):
+            # for/ site_idx in range(len(self.sites)):
+            #     site = self.sites[site_idx]
+                # Update progress on the milling sites left
+                start_time = time.time()
+                remaining_t = self.estimate_milling_time(site_idx, actual_time_per_site)
+                self._future.set_end_time(time.time() + remaining_t)
+
+                # .sites[idx].status already instantiated as FEATURE_ACTIVE
+                # Move the stage
+                target_pos = {"x": site.pos.value[0],
+                              "y": site.pos.value[1],
+                              "z": site.pos.value[2]}
+                # Todo check if it is wise to move in Z direction without affecting the OPTICAL-FIB alignment
+                logging.debug("For feature %s moving the stage to %s m", site.name.value, target_pos)
+                self._future.running_subf = self._stage.moveAbs(target_pos)
+                stage_time = self.estimate_stage_movement_time(site)
+                t = stage_time + 5  # adding extra margin
+                try:
+                    self._future.running_subf.result(t)  # or 2 min should be fast enough to move anywhere
+                except TimeoutError:
+                    logging.warning("Failed to move the stage for feature %s within %s s", site.name.value, t)
+                    self._future.running_subf.cancel()
+                    continue  # Try the next feature
+
```

Comment:
Please add a `move.cryoSwitchSamplePosition(MILLING)`, so that we are certain the hardware is in the right position (in case of the FM streams changing to FM_IMAGING). 
For now, you can leave it commented out.

Reply:
Should I cancel the sub_future in the except block? As it other sub-futures we are only using "raise" and logging commands.